engines/ultima/ultima8/graphics/raw_shape_frame.cpp

```diff
@@ -0,0 +1,162 @@
+#define READ2(data,offset) (data[offset] + (data[offset+1] << 8))
+#define READ4(data,offset) (data[offset] + (data[offset+1] << 8) + (data[offset+2] << 16) + (data[offset+3] << 24))
+
+// This will load a u8 style shape 'optimized'.
+void RawShapeFrame::LoadU8Format(const uint8 *data, uint32 /*size*/) {
+	_compressed = READ1(data, 8);
+	_width = static_cast<int16>(READ2(data, 10));
+	_height = static_cast<int16>(READ2(data, 12));
+	_xoff = static_cast<int16>(READ2(data, 14));
+	_yoff = static_cast<int16>(READ2(data, 16));
+
+	if (_height == 0)
+		return;
+
+	_line_offsets = new uint32[_height];
+
+	data += 18;
+	for (int32 i = 0; i < _height; i++) {
+		_line_offsets[i] = READ2(data, 0) - ((_height - i) * 2);
+		data += 2;
+	}
+
+	_rle_data = data;
+}
+
+// This will load a pentagram style shape 'optimzed'.
+void RawShapeFrame::LoadPentagramFormat(const uint8 *data, uint32 /*size*/) {
+	_compressed = READ1(data, 0);
+	_width = static_cast<int32>(READ4(data, 4));
+	_height = static_cast<int32>(READ4(data, 8));
+	_xoff = static_cast<int32>(READ4(data, 12));
+	_yoff = static_cast<int32>(READ4(data, 16));
+
+	if (_height == 0) return;
+
+	_line_offsets = new uint32[_height];
+
+	data += 20;
+	for (int32 i = 0; i < _height; i++) {
+		_line_offsets[i] = READ4(data, 0);
+		data += 4;
+	}
+
+	_rle_data = data;
+}
+
+// This will load any sort of shape via a ConvertShapeFormat struct
+void RawShapeFrame::LoadGenericFormat(const uint8 *data, uint32 size, const ConvertShapeFormat *format) {
+	IBufferDataSource ds(data, size);
+
+	ds.skip(format->_bytes_frame_unknown);
+	_compressed = ds.readX(format->_bytes_frame_compression);
+	_width = ds.readXS(format->_bytes_frame_width);
+	_height = ds.readXS(format->_bytes_frame_height);
+	_xoff = ds.readXS(format->_bytes_frame_xoff);
+	_yoff = ds.readXS(format->_bytes_frame_yoff);
+
+	if (_height == 0) return;
+
+	_line_offsets = new uint32[_height];
+
+	if (format->_line_offset_absolute) for (int32 i = 0; i < _height; i++) {
+			_line_offsets[i] = ds.readX(format->_bytes_line_offset);
+		}
+	else for (int32 i = 0; i < _height; i++) {
+			_line_offsets[i] = ds.readX(format->_bytes_line_offset) - ((_height - i) * format->_bytes_line_offset);
+		}
+
+	_rle_data = data + format->_len_frameheader2 + _height * format->_bytes_line_offset;
+}
+
+// This will load an U8-compressed shape
+void RawShapeFrame::LoadU8CMPFormat(const uint8 *data, uint32 size, const ConvertShapeFormat *format, const uint8 special[256], ConvertShapeFrame *prev) {
+	IBufferDataSource ds(data, size);
+
+	ConvertShapeFrame f;
+
+	f.ReadCmpFrame(&ds, format, special, prev);
+
+	uint32 to_alloc = f._height + (f._bytes_rle + 3) / 4;
+	_line_offsets = new uint32[to_alloc];
+	_rle_data = reinterpret_cast<uint8 *>(_line_offsets + f._height);
+
+	_compressed = f._compression;
+	_height = f._height;
+	_width = f._width;
+	_xoff = f._xoff;
+	_yoff = f._yoff;
+
+	Std::memcpy(_line_offsets, f._line_offsets, f._height * 4);
```

Comment:
Your `Std::memcpy` class is just a wrapper over `memcpy`. I'd drop that altogether.

Reply:
Yeah, also leftover that I should go through and clean all those up.  For now I have fixed this branch so it's not adding new ones at least.